Docker

iankronquist opened the pull request:
This is very preliminary. The Dockerfile works, but I need to add docs.
The biggest change is that I added a dependency, python-decouple, which was recommended by @mythmon as a way to use and environment variable based config. Mozilla uses it in a couple projects, and it looks like a clean solution.
I also changed the settings.py.dist to use decouple.
Sample docker commands might look like:

``` shell
$ docker build -t osl_test/pgd .
$ docker run --name pgd_mysql -e MYSQL_ROOT_PASSWORD=pgd_mysql -e MYSQL_USER=root -e MYSQL_PASSWORD=pgd_db -e MYSQL_DATABASE=pgd_db -d mysql
$ docker run --name pgd --link pgd_mysql:mysql osl_test/pgd
```

iankronquist committed:
Add python-decouple for environment based config

Optional of course. An environment based config makes dockerization much
easier.

iankronquist committed:
Use decouple for an optional environment config

By default it falls back to previous values.

iankronquist committed:
Add a Dockerfile! 🐳

iankronquist committed:
Change default environment variables

"root" is typically already a user on mysql. Chose a different name and
made the defaults more simple.

iankronquist commented:
@mathuin Alright, take a look at this.

mathuin commented:
Okay, checked out your branch on coreos1 and ran your sample forms, and got "access denied for user 'pgd'@'172.17.0.7' (using password:  YES)" which can probably be fixed by someone who knows more about the mysql container than I could glean from https://registry.hub.docker.com/_/mysql/.

Other than that, it looks pretty reasonable -- the settings.py changes will have to merge with the changes from the cookbook once those are done, but that's neither here nor there.  I have a comment to make on one line but I'll do that separately.

mathuin commented on settings.py.dist:
```diff
@@ -10,57 +12,57 @@ ADMINS = (
 
 #Google analytics ID. Enter the full id, as in: UA-xxxxxx-x 
-GOOGLE_ID = 'UA-9880071-1'
+GOOGLE_ID = config('GOOGLE_ID', default='UA-9880071-1')
```
Use a default value of 'None' instead of that string, since that's our official string and should not be in the repo. :-)

mathuin commented on settings.py.dist:
```diff
@@ -10,57 +12,57 @@ ADMINS = (
 
 # Make this unique, and don't share it with anybody.
-SECRET_KEY = 'g3#f517j(b4$ze*3_8pi3i$=1@mk=fnbb_gunjwk_f8whrbo=a'
+SECRET_KEY = config('SECRET_KEY', default='g3#f517j(b4$ze*3_8pi3i$=1@mk=fnbb_gunjwk_f8whrbo=a')
```
This should also not be in the repository.  Is there a clever way to make Django generate a new one if this one is set to '' or 'None' or something?

chancez replied:
Generating one is difficult to automate, because it needs to be the same between runs for some apps. It would need to:
- Generate a key
- Write key to file
- Use key from file in subsequent runs

ramereth replied:
Ideally you would set this manually in an encrypted databag for each instance. For testing purposes we can easily add an encrypted data bag with a bogus key. Might be useful to add in the docs how to generate one.

mathuin commented:
I worked with @ecnahc515 on this a while and we were able to make an environment file containing database parameters.  I'm not convinced we've got it just right, but I am way excited about the progress.

mathuin commented:
What do you think about documenting the need for a container with link name mysql ("--link pgd_mysql:mysql" etc) and then changing the environment variable DATABASE_HOST and DATABASE_PORT to something like MYSQL_PORT_3306_TCP_ADDR and MYSQL_PORT_3306_TCP_PORT ?  That would remove the need for the "magic values" currently in place.

iankronquist committed:
Add a fig file to make docker easier

Some devs have had trouble with long docker incantations before. This makes it
easier on them.

iankronquist committed:
Add fig to dependencies

iankronquist committed:
Add documentation on how to use docker

mathuin committed:
Minor code-block corrections plus one grammatical change

The code blocks weren't rendering so I read up a little on RST and
fixed them.  I also filled in the last bit which seemed a little
incomplete.  Still working on the code part of things!

mathuin committed:
Used container-specific variables in database configuration.

The PGD requires the use of a MySQL database at this time.  The
settings.py has been changed to rely upon a container with the link
name 'mysql' which runs a MySQL daemon.  If this works, I will change
the documentation to support this approach.

mathuin committed:
Updated Dockerfile to use best practices

Among Docker's documentation pages is an article on the best practices
for writing Dockerfiles.  In accordance with this document, I have
updated our Docker configuration in the following ways: added a
.dockerignore file which currently only contains .git; sorted
multi-line arguments, and attempted to use ADD versus COPY correctly.
The resulting Dockerfile looks great and actually works!

mathuin committed:
Updated documentation and fig.yml to match, plus minor changes.

Set the MYSQL_* variables in docs and fig.yml to more reasonable
values which match.  Also noticed a grammatical error in the
documentation and fixed it, and removed the reference to volumes in
the fig.yml which kept it from working.  For what it's worth, I think
this means the volumes suggestion at the foot of the documentation
page may not work as expected, since it mounts the current checkout
over /opt/pgd thus wiping out the settings.py file.

mathuin committed:
Set DOC_ROOT to the work directory.

Older Django (like the version we run) has a DOC_ROOT variable which
in our case should be set to the project's root directory.  I added
some boilerplate code to set PROJECT_ROOT to the project root
directory, then set DOC_ROOT to PROJECT_ROOT.  When we update Django,
this will need to be revisited.

mathuin committed:
Updated settings.py.dist to support database tests

The pgd_splicer tests require foreign keys to be disabled, which
requires a tweak to the database options variable.  This change has
been made on zeus but not in the repository so it wasn't found
earlier.

mathuin commented:
Checking out this branch and running "fig up" will create a MySQL database container and a Django web container running the PGD app.  A good start!

The following steps are required to demonstrate the PGD app:
 * add the OSUOSL yum repository for Centos 6 x86_64 and install the osuosl-dssp package.  The cookbook has this functionality.
 * create the necessary database tables with the following command:
````
$ fig run web python manage.py syncdb
````
I think this is in the cookbook too.
 * either create a small selection file as a subset of an existing one (say the first hundred lines) or if you want to demo the full database, use this command: 
````
fig run web python ./pgd_splicer/dunbrack_selector.py > selection.txt
````
This command and the two that follow do not generate any visible output.  This should be fixed.
 * retrieve the protein files from the website with this command:
````
fig run web python ./pgd_splicer/ftpupdate.py --pipein < selection.txt
````
Again, not seeing any output so it makes me a little nervous.
 * import the proteins into the database with this command:
````
fig run web python ./pgd_splicer/ProcessPDBTask.py --pipein < selection.txt
````
This command can take a long time!

Now spin up a web browser on the machine running the docker container and connect to http://localhost:8000/ and you should see the PGD.  Click 'search' and perform the default search, and you should get output similar to that seen in production.  Yay!

The one thing that is really needed before this can be used for active development is the OSUOSL repo and dssp binary stuff.  Once that's done, the output issue can be worked around by executing bash instead of the python commands, and then running the python commands within that bash shell.

iankronquist commented:
Looks pretty good, except for that disconcerting lack of output. You can get some insight into what's going on with `fig logs`, but it's not what you want.
Also, the MySQL server is showing a warning:
```
150213 18:31:59 [Note] Plugin 'FEDERATED' is disabled.
150213 18:31:59 InnoDB: The InnoDB memory heap is disabled
150213 18:31:59 InnoDB: Mutexes and rw_locks use GCC atomic builtins
150213 18:31:59 InnoDB: Compressed tables use zlib 1.2.3
150213 18:31:59 InnoDB: Using Linux native AIO
150213 18:31:59 InnoDB: Initializing buffer pool, size = 128.0M
150213 18:31:59 InnoDB: Completed initialization of buffer pool
150213 18:31:59 InnoDB: highest supported file format is Barracuda.
150213 18:31:59  InnoDB: Waiting for the background threads to start
150213 18:32:00 InnoDB: 5.5.42 started; log sequence number 1595675
150213 18:32:00 [Note] Server hostname (bind-address): '0.0.0.0'; port: 3306
150213 18:32:00 [Note]   - '0.0.0.0' resolves to '0.0.0.0';
150213 18:32:00 [Note] Server socket created on IP: '0.0.0.0'.
150213 18:32:00 [Note] Event Scheduler: Loaded 0 events
150213 18:32:00 [Note] mysqld: ready for connections.
Version: '5.5.42'  socket: '/tmp/mysql.sock'  port: 3306  MySQL Community Server (GPL)
150213 18:32:01 [Warning] IP address '172.17.0.129' could not be resolved: Name or service not known
150213 18:32:23 [Warning] IP address '172.17.0.130' could not be resolved: Name or service not known
```

mathuin commented:
The stuff before the Version line is the price of using the stock container, and in general I'm okay with that.

The last two lines are representative of a well-known issue -- http://www.percona.com/blog/2008/05/31/dns-achilles-heel-mysql-installation/ is a decent description of the problem and one way to address it.  That being said, it's relatively harmless and lets us use the stock container without changes and I kinda like that.  I do not care strongly enough to argue either way, though: as long as the container continues to work, I'm satisfied.

iankronquist commented:
Fair enough. I'd say that it looks good to me and deserves an +1

mathuin commented:
@iankronquist I revisited this pull request today.  Can you add the OSUOSL repo and the dssp binary?  Once that's done, I can develop some additional documentation based on my earlier comment.  Thanks!

mathuin committed:
Added OSUOSL's DSSP RPM to the build

There is no existing file for us to download, and I didn't think it
was appropriate to add the file to the repository.  Docker doesn't
support here documents so I build the file line by line.  The next
step is to document how to use Docker with PGD.

mathuin commented:
Okay, I just added the repo for the binary.  I will write more docs this week, and then I'll ask for review.  Thanks for being patient!

mathuin committed:
Minor additions to documentation

I clarified some minor things in the docs with respect to fig, and
what actually happens when some commands are run.

mathuin commented:
@muricula I added some more to the docs, but I would like you to run through all the examples in the Docker section of the docs (to confirm that they all work as expected) before I add the stuff from https://github.com/osuosl/pgd/pull/8#issuecomment-74354539.  After that, I think we're just about done.

iankronquist committed:
Add instructions on how to connect to the app

iankronquist committed:
Add notes about fig

Note that race conditions happen.
Note that we should rename fig to docker-compose

iankronquist commented:
@mathuin All the examples worked for me. I added a few things.

mathuin committed:
Added quick start to documentation.

Now that fig and docker are working as expected, the documentation
needed a "quick start" to help potential developers and testers spin
up a small (100-protein) instance of the PGD.  This quick start
assumes familiarity with the new tools, but there are sufficient
checks along the way to confirm that progress is being made.

mathuin committed:
Updated code and instructions to use pdb container

A data-only container was added to the configuration file and the
documentation.  This container provides persistent space for
downloaded PDB files.

mathuin committed:
Added setup and test scripts

Two scripts were added to support the new Docker configuration.  One
script sets up the containers as described in the documentation, and
the other test performs an integration test on those containers.

mathuin committed:
Modified docs to reflect new scripts

The new Docker documentation includes references to the two new
testing scripts.

iankronquist commented:
Looks good to me.
